synthtool/sources/git.py

```diff
@@ -141,3 +149,29 @@ def extract_commit_message_metadata(message: str) -> Dict[str, str]:
         metadata[key] = value.strip()
 
     return metadata
+
+
+PRECLONE_MAP_HELP = """
+A preclone map file contains a list of git repos that have already been cloned.
+Its contents are json and look like this:
+{
+  "http://github.com/googleapis/nodejs-vision.git": "/tmp/nodejs-vison",
+  "http://github.com/googleapis/googleapis": "/tmp/googleapis"
+}
+"""
+
+
+@functools.lru_cache(maxsize=None)
+def _get_preclone_map() -> Dict:
+    """Loads the preclone map from the file specified in an environment variable."""
+    preclone_map_path = os.environ.get(PRECLONE_MAP_ENVIRONMENT_VARIABLE)
+    if not preclone_map_path:
+        return {}
+    with open(preclone_map_path, "r") as json_file:
+        return json.load(json_file)
+
+
+def get_preclone(url: str) -> Optional[str]:
+    """Finds a pre-cloned git repo in the preclone map."""
+    preclone_map = _get_preclone_map()
+    return preclone_map.get(url)
```

Comment:
Do we want to try and be lenient on the format of the git url? i.e. `git@github.com:owner/repo.git` vs `https://github.com/owner/repo` vs `https://github.com/owner/repo.git`

Reply:
The plan is that autosynth alone will supply a preconfig.json file.  It will populate the preconfig file by looking at the previous synthtool's output synthtool.metadata file.  Therefore the urls should match exactly.  Do you see another use case or an opportunity for mismatch?